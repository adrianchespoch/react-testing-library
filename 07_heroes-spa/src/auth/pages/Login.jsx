import { useNavigate } from 'react-router-dom';
import { useAuth } from '../../hooks/useAuth';

export const Login = () => {
  const navigate = useNavigate();

  const { login } = useAuth();

  const handleLogin = () => {
    const lastPath = localStorage.getItem('lastPath') || '/';

    login('Alex Auth');

    navigate(lastPath, { replace: true });
  };

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />

      <button onClick={handleLogin} className="btn btn-primary">
        Login
      </button>
    </div>
  );
};
