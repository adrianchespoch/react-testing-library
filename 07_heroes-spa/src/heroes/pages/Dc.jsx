import { HeroList } from '../components';

export const Dc = () => {
  return (
    <>
      <h1>DC Comics</h1>

      <HeroList publisher={'DC Comics'} />
    </>
  );
};
