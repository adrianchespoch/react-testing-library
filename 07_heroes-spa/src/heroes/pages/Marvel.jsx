import { HeroList } from '../components';

export const Marvel = () => {
  return (
    <>
      <h1>Marvel Comics</h1>

      <HeroList publisher={'Marvel Comics'} />
    </>
  );
};
