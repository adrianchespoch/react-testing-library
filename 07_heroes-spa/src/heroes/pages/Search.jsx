import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../components';
import { getHeroesByName } from '../helpers';

export const Search = () => {
  const navigate = useNavigate();
  const location = useLocation();

  // Procesar las query params <- x si viene >1
  const { q = '' } = queryString.parse(location.search);

  const { searchText, onInputChange } = useForm({ searchText: q });

  const heroes = getHeroesByName(q);

  const handleSubmit = e => {
    e.preventDefault();
    if (!searchText.trim()) return;

    // Enviar los query params: Agregar los QP a la url sin refresh
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />

          <form onSubmit={handleSubmit} role="searchForm" data-testid="searchForm">
            <input
              type="text"
              placeholder="Search a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />

            <button className="btn btn-primary mt-3">Search</button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {!q ? (
            <div className="alert alert-info" aria-label="search">
              Search for an hero
            </div>
          ) : (
            !heroes.length && (
              <div aria-label="nothing" className="alert alert-danger">
                Nothing to do: {q}
              </div>
            )
          )}

          {heroes.map(hero => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
