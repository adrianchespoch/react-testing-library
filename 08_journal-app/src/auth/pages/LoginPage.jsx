import { useMemo } from 'react';
import { Link as RouterLink } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import {
  Alert,
  Button,
  Grid,
  Link,
  TextField,
  Typography,
} from '@mui/material';
import { Google } from '@mui/icons-material';

import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from './../../hooks';
import {
  startGoogleSignIng,
  startLoginWithEmailAndPass,
} from '../../store/auth';

const initState = {
  email: '',
  password: '',
};

export const LoginPage = () => {
  const dispatch = useDispatch();
  const { status, errorMessage } = useSelector(state => state.auth);

  const { email, password, handleInputChange } = useForm(initState);

  const isAuthenticating = useMemo(() => status === 'checking', [status]);

  const handleSubmit = e => {
    e.preventDefault();
    if (!email || !password) return;

    dispatch(startLoginWithEmailAndPass(email, password));
  };

  const onGoogleSignIn = () => {
    dispatch(startGoogleSignIng());
  };

  return (
    <AuthLayout title="Login">
      <form
        onSubmit={handleSubmit}
        className="animate__animated animate__fadeIn animate__faster"
        data-testid="form"
      >
        <Grid container>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Email"
              type="email"
              placeholder="test@test.com"
              fullWidth
              name="email"
              value={email}
              onChange={handleInputChange}
            />
          </Grid>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Password"
              type="password"
              placeholder="Password"
              fullWidth
              name="password"
              inputProps={{ 'data-testid': 'password' }}
              value={password}
              onChange={handleInputChange}
            />
          </Grid>

          <Grid container display={errorMessage ? '' : 'none'} sx={{ mt: 2 }}>
            <Grid item xs={12}>
              <Alert severity="error">{errorMessage}</Alert>
            </Grid>
          </Grid>

          <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
            <Grid item xs={12} sm={6}>
              <Button
                type="submit"
                variant="contained"
                fullWidth
                disabled={isAuthenticating}
              >
                Login
              </Button>
            </Grid>
            <Grid item xs={12} sm={6}>
              <Button
                variant="contained"
                fullWidth
                onClick={onGoogleSignIn}
                disabled={isAuthenticating}
                aria-label="google-btn"
              >
                <Google />
                <Typography sx={{ ml: 1 }}>Google</Typography>
              </Button>
            </Grid>
          </Grid>

          <Grid container direction="row" justifyContent="end">
            <Link component={RouterLink} color="inherit" to="/auth/register">
              Crear una cuenta
            </Link>
          </Grid>
        </Grid>
      </form>
    </AuthLayout>
  );
};
