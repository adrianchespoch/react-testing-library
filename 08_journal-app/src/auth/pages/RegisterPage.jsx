import { useMemo, useState } from 'react';
import { Link as RouterLink } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import {
  Button,
  Grid,
  TextField,
  Typography,
  Link,
  Alert,
} from '@mui/material';

import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks';
import { startCreatingUserWithEmailPassword } from '../../store/auth';

const initFormData = {
  displayName: '',
  email: '',
  password: '',
};

const formValidations = {
  email: [value => value.includes('@'), 'El correo debe tener una @'],
  password: [
    value => value.length >= 6,
    'El password debe tener mas de 6 caracteres!',
  ],
  displayName: [value => value.length >= 1, 'El nombre es obligatorio'],
};

export const RegisterPage = () => {
  const dispatch = useDispatch();
  const [formSubmited, setFormSubmited] = useState(false);

  const { status, errorMessage } = useSelector(state => state.auth);
  const isCheckingAuth = useMemo(() => {
    status === 'checking';
  }, [status]);

  const {
    displayName,
    email,
    password,
    handleInputChange,
    formValues,
    displayNameValid,
    emailValid,
    passwordValid,
    isFormValid,
  } = useForm(initFormData, formValidations);

  const handleSubmit = e => {
    e.preventDefault();

    // Form validations
    setFormSubmited(true);
    if (!isFormValid) return;

    // Peticion a firebase para crear user con email & pass
    dispatch(startCreatingUserWithEmailPassword(formValues));
  };

  return (
    <AuthLayout title="Login">
      <form onSubmit={handleSubmit}>
        <h1>FormValid: {isFormValid ? 'Valid' : 'Invalid'}</h1>

        <Grid container>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Name"
              type="text"
              placeholder="John Doe"
              fullWidth
              name="displayName"
              value={displayName}
              onChange={handleInputChange}
              error={!!displayNameValid && formSubmited}
              helperText={formSubmited && displayNameValid}
            />
          </Grid>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Email"
              type="email"
              placeholder="test@test.com"
              fullWidth
              name="email"
              value={email}
              onChange={handleInputChange}
              error={!!emailValid && formSubmited}
              helperText={formSubmited && emailValid}
            />
          </Grid>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Password"
              type="password"
              placeholder="Password"
              fullWidth
              name="password"
              value={password}
              onChange={handleInputChange}
              error={!!passwordValid && formSubmited}
              helperText={formSubmited && passwordValid}
            />
          </Grid>

          <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
            {/* {errorMessage && (
              <Grid item xs={12}>
                <Alert severity="error">{errorMessage}</Alert>
              </Grid>
            )} */}
            {/* Testing:  */}
            <Grid item xs={12} display={errorMessage ? '' : 'none'}>
              <Alert severity="error">{errorMessage}</Alert>
            </Grid>

            <Grid item xs={12}>
              <Button
                variant="contained"
                fullWidth
                type="submit"
                disabled={isCheckingAuth}
              >
                Crear cuenta
              </Button>
            </Grid>
          </Grid>

          <Grid container direction="row" justifyContent="end">
            <Typography sx={{ mr: 1 }}>Ya tienes cuenta?</Typography>
            <Link component={RouterLink} color="inherit" to="/auth/login">
              Ingresar
            </Link>
          </Grid>
        </Grid>
      </form>
    </AuthLayout>
  );
};
