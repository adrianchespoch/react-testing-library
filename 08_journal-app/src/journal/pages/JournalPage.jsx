import { useDispatch, useSelector } from 'react-redux';
import { AddOutlined } from '@mui/icons-material';
import { IconButton } from '@mui/material';

import { JournalLayout } from '../layout/JournalLayout';
import { NothingSelectedView, NoteView } from '../views';
import { startNewNote } from '../../store/journal';

export const JournalPage = () => {
  const dispatch = useDispatch();
  const { isSaving, activeNote } = useSelector(state => state.journal);

  const handleNewNote = () => {
    dispatch(startNewNote());
  };

  return (
    <JournalLayout>
      {activeNote?.id ? <NoteView /> : <NothingSelectedView />}
      {/* {!!activeNote ? <NoteView /> : <NothingSelectedView />} */}

      <IconButton
        size="large"
        sx={{
          color: 'white',
          backgroundColor: 'error.main',
          ':hover': { backgroundColor: 'error.main', opacity: 0.9 },
          position: 'fixed',
          right: 50,
          bottom: 50,
        }}
        onClick={handleNewNote}
        disabled={isSaving}
      >
        <AddOutlined sx={{ fontSize: 30 }} />
      </IconButton>
    </JournalLayout>
  );
};
