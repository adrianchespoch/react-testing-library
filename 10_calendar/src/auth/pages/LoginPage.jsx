import { useEffect } from 'react';
import Swal from 'sweetalert2';
import { useAuthStore, useForm } from '../../hooks';
import './LoginPage.css';

const loginFormFields = {
  loginEmail: '',
  loginPassword: '',
};

const registerFormFields = {
  registerName: '',
  registerEmail: '',
  registerPassword: '',
  registerPassword2: '',
};

export const LoginPage = () => {
  const { startLogin, errorMessage, startRegister } = useAuthStore();
  const {
    loginEmail,
    loginPassword,
    handleInputChange: onLoginInputChange,
    formValues: lformValues,
  } = useForm(loginFormFields);

  const {
    registerName,
    registerEmail,
    registerPassword,
    registerPassword2,
    handleInputChange: onRegisterInputChange,
    formValues: rformValues,
  } = useForm(registerFormFields);

  useEffect(() => {
    errorMessage &&
      Swal.fire(errorMessage, 'Revisar las credenciales', 'error');
  }, [errorMessage]);

  const handleLoginSubmit = async e => {
    e.preventDefault();
    if (Object.values(lformValues).some(field => !field)) return;

    await startLogin({ email: loginEmail, password: loginPassword });
  };

  const handleRegisterSubmit = async e => {
    e.preventDefault();
    if (Object.values(rformValues).some(field => !field)) return;

    await startRegister({
      name: registerName,
      email: registerEmail,
      password: registerPassword,
    });
  };

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Ingreso</h3>
          <form onSubmit={handleLoginSubmit}>
            <div className="form-group mb-2">
              <input
                type="text"
                className="form-control"
                placeholder="Correo"
                name="loginEmail"
                value={loginEmail}
                onChange={onLoginInputChange}
              />
            </div>
            <div className="form-group mb-2">
              <input
                type="password"
                className="form-control"
                placeholder="Contraseña"
                name="loginPassword"
                value={loginPassword}
                onChange={onLoginInputChange}
              />
            </div>
            <div className="form-group mb-2">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Registro</h3>
          <form onSubmit={handleRegisterSubmit}>
            <div className="form-group mb-2">
              <input
                type="text"
                className="form-control"
                placeholder="Nombre"
                name="registerName"
                value={registerName}
                onChange={onRegisterInputChange}
              />
            </div>
            <div className="form-group mb-2">
              <input
                type="email"
                className="form-control"
                placeholder="Correo"
                name="registerEmail"
                value={registerEmail}
                onChange={onRegisterInputChange}
              />
            </div>
            <div className="form-group mb-2">
              <input
                type="password"
                className="form-control"
                placeholder="Contraseña"
                name="registerPassword"
                value={registerPassword}
                onChange={onRegisterInputChange}
              />
            </div>

            <div className="form-group mb-2">
              <input
                type="password"
                className="form-control"
                placeholder="Repita la contraseña"
                name="registerPassword2"
                value={registerPassword2}
                onChange={onRegisterInputChange}
              />
            </div>

            <div className="form-group mb-2">
              <input type="submit" className="btnSubmit" value="Crear cuenta" />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
