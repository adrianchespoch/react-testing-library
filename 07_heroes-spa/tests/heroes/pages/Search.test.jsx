import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { Search } from '../../../src/heroes/pages/Search';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

describe('Pruebas en <Navbar />', () => {
  beforeEach(() => jest.clearAllMocks());

  test('debe mostrar el nombre del user auth', () => {
    const { container } = render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test('debe renderizar a Batman y el input con el valor del queryString', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Search />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    expect(input.value).toBe('batman');

    const img = screen.getByRole('img');
    expect(img.src).toContain('/assets/heroes/dc-batman.jpg');

    expect(screen.getByText('Batman')).toBeTruthy();
    expect(screen.getByText('Bruce Wayne')).toBeTruthy();

    // // No funca
    // expect(screen.getByLabelText('search')).not.toBeInTheDocument();

    /* // // Si se renderizara el alert con  style: display: none
    const alert = screen.getByLabelText('alert-danger');
    expect(alert.style.display).toBe('none');
     */
  });

  test('debe monstrar el error si NO se encuentra el hero', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=xyzd']}>
        <Search />
      </MemoryRouter>
    );

    expect(screen.getByLabelText('nothing')).toBeTruthy();
    expect(screen.getByLabelText('nothing').innerHTML).toBe(
      'Nothing to do: xyzd'
    );
  });

  test('debe llamar el navigate con los params correctos', () => {
    const inputValue = 'green';

    render(
      <MemoryRouter initialEntries={['/search']}>
        <Search />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');

    // const form = screen.getByRole('searchForm');     // con un  role=''
    const form = screen.getByTestId('searchForm');     // con un   data-testid=''

    fireEvent.change(input, {
      target: { name: 'searchText', value: inputValue },
    });

    fireEvent.submit(form);

    expect(mockedUseNavigate).toHaveBeenCalled();
    expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
  });
});
