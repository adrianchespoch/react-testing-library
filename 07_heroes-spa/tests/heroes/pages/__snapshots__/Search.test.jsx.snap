// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Navbar /> debe mostrar el nombre del user auth 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        data-testid="searchForm"
        role="searchForm"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-3"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="search"
        class="alert alert-info"
      >
        Search for an hero
      </div>
    </div>
  </div>
</div>
`;
